### PR TITLE
Fb billing

### PR DESCRIPTION
**Pull request by Javasamurai**

### CODE REVIEW

File: js/index.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import .browser;
```
Comment (rampr) [inline js/index.js:1]: Is this okay to include browser always ?

Comment (Javasamurai) [reply js/index.js:1]: Not necessary for platforms apart from web

**Comment by rampr** [comment]
Isn't the fb billing api going to be in the facebook module ?

**Comment by Javasamurai** [comment]
I think it's consistent to put it in billing module. @rampr

File: js/index.js [hunk 2]
```diff
@@ -236,7 +237,7 @@ var Billing = Class(Emitter, function (supr) {
   this.purchase = simulatePurchase;
 });
 
-var billing = new Billing;
+var billing = new Billing();
```
Comment (rampr) [inline js/index.js:240]: is this needed ?

File: js/browser.js [hunk 3]
```diff
@@ -8,25 +8,25 @@ var Billing = Class(Emitter, function (supr) {
...
-		if (data.error_code) {
+		if (!data || data.error_code) {
 			if (typeof this.onFailure === 'function') {
 				logger.info("BILLING : onFailure of Billing plugin");
 				this.onFailure(data);
 			} else {
 				logger.info("BILLING : onFailure of Billing plugin is not defined");
 			}
-		} else if (data.status === 'completed') {
+		} else if (data && data.status === 'completed') {
```
Comment (vivekb91) [inline js/browser.js]: we are already checking data line no 15.

File: js/browser.js [hunk 4]
```diff
@@ -8,25 +8,25 @@ var Billing = Class(Emitter, function (supr) {
 			product : item.product, // URL of open graph object browser only
 			quantity: item.quantity,
 			request_id: item.reciept
-		}, this.callback);
+		}, bind(this, this.callback));
```
Comment (rampr) [inline js/browser.js]: this may not be necessary

File: js/index.js (hunk 1 above)
Comment (rampr) [reply js/index.js:1]: How can we exclude it for platforms apart from web ?

File: js/browser.js [hunk 5]
```diff
@@ -4,32 +4,36 @@ var Billing = Class(Emitter, function (supr) {
...
 			method: 'pay',
-			action: 'purchaseitem',
-			product : item.product, // URL of open graph object browser only
+			action: 'purchaseiap',
+			product_id: item,
 			quantity: item.quantity,
-			request_id: item.reciept
-		}, this.callback);
+			request_id: item.reciept,
+			developer_payload: 'Boom boom pow!'
```
Comment (rampr) [inline js/browser.js]: huh ?

Comment (Javasamurai) [reply js/browser.js]: Just for testing! 😃

File: js/index.js (hunk 2 above)
Comment (Javasamurai) [reply js/index.js:240]: Both works the same. Not necessary though

File: js/index.js (hunk 1 above)
Comment (rampr) [reply js/index.js:1]: As discussed, it isn't needed

File: js/browser.js [hunk 6]
```diff
@@ -4,32 +4,35 @@ var Billing = Class(Emitter, function (supr) {
 	this.purchase = function (item) {
```
Comment (rampr) [inline js/browser.js]: Can you convert this file to 2 spaces. Other files in this project use 2 spaces

File: js/browser.js [hunk 7]
```diff
@@ -4,32 +4,35 @@ var Billing = Class(Emitter, function (supr) {
...
 		FB.ui({
 			method: 'pay',
-			action: 'purchaseitem',
-			product : item.product, // URL of open graph object browser only
+			action: 'purchaseiap',
+			product_id: item,
 			quantity: item.quantity,
 			request_id: item.reciept
-		}, this.callback);
+		}, bind(this, function (data) {
```
Comment (rampr) [inline js/browser.js]: This might work

```
FB.ui({
...
}, bind(this, this.callback, item));
```

File: README.md [hunk 8]
```diff
@@ -15,6 +15,9 @@ You can import the billing object anywhere in your application:
 import billing as billing;
 ~~~
 
+### facebook web Setup
+
+For facebook iap to integrate with your game make sure the `product id` is same in both app dashboard and game.
```
Comment (rampr) [inline README.md]: Please add steps on how to configure IAP in the game

File: js/browser.js (hunk 6 above)
Comment (Javasamurai) [reply js/browser.js]: Okay.

File: README.md (hunk 8 above)
Comment (Javasamurai) [reply README.md]: We don't have to @rampr as there is no change for facebook or device. Only product id needs to be same.

File: js/browser.js [hunk 9]
```diff
@@ -1,40 +1,43 @@
...
-		FB.ui({
-			method: 'pay',
-			action: 'purchaseitem',
-			product : item.product, // URL of open graph object browser only
-			quantity: item.quantity,
-			request_id: item.reciept
-		}, this.callback);
-	};
+  this.purchase = function (item) {
+	FB.ui({
```
Comment (rampr) [inline js/browser.js]: still using tab here

File: js/browser.js [hunk 10]
```diff
@@ -1,40 +1,43 @@
...
-		} else if (data.status === 'initiated') {
-			if (typeof this.onPurchase === 'function') {
-				logger.info("BILLING : < async > onPurchase of Billing plugin");
-				this.onPurchase(data, true);
-			} else {
-				logger.info("BILLING : < async > onPurchase of Billing plugin is not defined");
-			}
+  this.callback = function (data, item) {
+    if (!data || data.error_code) {
+	  if (typeof this.onFailure === 'function') {
```
Comment (rampr) [inline js/browser.js]: tab here too

File: js/browser.js [hunk 11]
```diff
@@ -1,40 +1,43 @@
...
+	  	logger.info("BILLING : onFailure of Billing plugin");
+	  	this.onFailure(item);
+	  } else {
+	  	logger.info("BILLING : onFailure of Billing plugin is not defined");
+	  }
+    } else if (data.status === 'completed') {
+		if (typeof this.onPurchase === 'function') {
```
Comment (rampr) [inline js/browser.js]: indentation issue here

File: js/browser.js [hunk 12]
```diff
@@ -1,40 +1,41 @@
...
-			product : item.product, // URL of open graph object browser only
-			quantity: item.quantity,
-			request_id: item.reciept
-		}, this.callback);
-	};
+  this.purchase = function (item) {
+    FB.ui({
+      method: 'pay',
+      action: 'purchaseiap',
+      product_id: item,
```
Comment (vivekb91) [inline js/browser.js]: why product id is item

File: js/browser.js [hunk 13]
```diff
@@ -1,40 +1,41 @@
...
+      quantity: item.quantity
+    }, bind(this, function (data) {
```
Comment (rampr) [inline js/browser.js]: `bind(this, this.callback, item);`
doesn't work ?

File: js/browser.js [hunk 14]
```diff
@@ -1,40 +1,41 @@
...
+    if (!data || data.error_code) {
+      if (typeof this.onFailure === 'function') {
+        logger.info("BILLING : onFailure of Billing plugin");
+        this.onFailure(item);
+      } else {
+        logger.info("BILLING : onFailure of Billing plugin is not defined");
+      }
+    } else if (data.status === 'completed') {
+        if (typeof this.onPurchase === 'function') {
+            logger.info("BILLING : < sync > onPurchase of Billing plugin");
```
Comment (rampr) [inline js/browser.js]: indentation wrong. Can you recheck everything properly.

File: js/browser.js [hunk 15]
```diff
@@ -1,40 +1,41 @@
...
+            this.onPurchase(item, false);
+        } else {
+            logger.info("BILLING : < sync > onPurchase of Billing plugin is not defined");
+        }
+    } else if (data.status === 'initiated') {
+        if (typeof this.onPurchase === 'function') {
+          logger.info("BILLING : < async > onPurchase of Billing plugin");
+          this.onPurchase(item, true);
+        } else {
+          logger.info("BILLING : < async > onPurchase of Billing plugin is not defined");
```
Comment (rampr) [inline js/browser.js]: what happens during async purchases ?
How does the application get notified ?

Comment (rampr) [reply js/browser.js]: Has this been tested ?

**Comment by rampr** [comment]
Can you list the test scenarios that were used ?

**Review by vivekb91 — APPROVED** [review]

File: js/browser.js [hunk 16]
```diff
@@ -1,40 +1,59 @@
...
-				this.onPurchase(data, true);
-			} else {
-				logger.info("BILLING : < async > onPurchase of Billing plugin is not defined");
-			}
-		}
-	};
+  var consumedItems = {};
+  var purchasedItems = {};
+
+
```
Comment (rampr) [inline js/browser.js]: unnecessary new line

**Review by rampr — APPROVED** [review]

**Comment by vivekb91** [comment]
Approved

**Pull request merged**